### PR TITLE
add binomial entropy and kl

### PR DESCRIPTION
**Pull request by alicanb**
This is a larger PR than I intended but basically it adds binomial entropy and binomial-poisson and binomial-geometric KL with some helper functions:
 - `binomial._log1pmprobs`: I used this a lot so I made it a separate function. it calculates
`(-probs).log1p()` safely.
-  `binomial._Elnchoosek()`: for x~Bin(n, p), this calculates E[log(nchoosek)], E[log(n!)], E[log(x!)], E[log((n-x)!)]

### CODE REVIEW

File: torch/distributions/binomial.py [hunk 1]
```diff
@@ -77,6 +77,11 @@ def probs(self):
     def param_shape(self):
         return self._param.size()
 
+    def _log1pmprobs(self):
```
Comment (vishwakftw) [inline torch/distributions/binomial.py]: Since it is a function for internal use, I think this can be moved to the top, like in MVN. Something like:

```python
def _log1pmtensor(tensor):
    # Do the same thing
```

Uses of the function in `kl.py` can be done via importing this function along with `Binomial`.

File: torch/distributions/binomial.py [hunk 2]
```diff
@@ -109,3 +111,27 @@ def enumerate_support(self):
         values = values.view((-1,) + (1,) * len(self._batch_shape))
         values = values.expand((-1,) + self._batch_shape)
         return values
+
+    def _Elnchoosek(self):
```
Comment (vishwakftw) [inline torch/distributions/binomial.py]: Same idea here.

File: torch/distributions/binomial.py [hunk 3]
```diff
@@ -109,3 +111,27 @@ def enumerate_support(self):
...
+        # return expected value of log(nchoosek), log(n!),log(k!), log(n-k!)
+        # where k~Bin(n,p)
+        s = self.enumerate_support()
+        s[0] = 1  # 0! = 1
+        # x is factorial matrix i.e. x[k,...] = k!
+        x = torch.cumsum(s.log(), dim=0)
```
Comment (vishwakftw) [inline torch/distributions/binomial.py]: `x` is the log of factorial matrix right?

File: torch/distributions/binomial.py [hunk 4]
```diff
@@ -109,3 +111,27 @@ def enumerate_support(self):
...
+        s[0] = 0
+        indices = [slice(None)] * x.dim()
+        indices[0] = torch.arange(x.size(0) - 1, -1, -1,
+                                  dtype=torch.long, device=x.device)
+        # x[tuple(indices)] is x reversed on first axis
+        lnchoosek = x[-1] - x - x[tuple(indices)]
```
Comment (vishwakftw) [inline torch/distributions/binomial.py]: I think `x.flip(dim=0)` will exhibit same behaviour.

File: torch/distributions/binomial.py [hunk 5]
```diff
@@ -109,3 +111,27 @@ def enumerate_support(self):
...
+        elognfac = x[-1]
+        elogkfac = ((lnchoosek + s * self.logits + self.total_count * self._log1pmprobs()).exp() *
+                    x).sum(dim=0)
+        elognmkfac = ((lnchoosek + s * self.logits + self.total_count * self._log1pmprobs()).exp() *
```
Comment (vishwakftw) [inline torch/distributions/binomial.py]: E[log(n-k)!] = E[log k!] but for Bin(n, (1 - p)). Can we use this fact here?

File: torch/distributions/kl.py [hunk 6]
```diff
@@ -199,12 +199,29 @@ def _kl_binomial_binomial(p, q):
...
     if (p.total_count < q.total_count).any():
         raise NotImplementedError('KL between Binomials where q.total_count > p.total_count is not implemented')
-    kl = p.total_count * (p.probs * (p.logits - q.logits) + (-p.probs).log1p() - (-q.probs).log1p())
+    kl = p.total_count * (p.probs * (p.logits - q.logits) + p._log1pmprobs() - q._log1pmprobs())
     inf_idxs = p.total_count > q.total_count
     kl[inf_idxs] = _infinite_like(kl[inf_idxs])
     return kl
 
 
+@register_kl(Binomial, Poisson)
```
Comment (vishwakftw) [inline torch/distributions/kl.py]: Heterogeneous combinations were placed below. This section was for homogeneous combinations.

File: torch/distributions/kl.py [hunk 7]
```diff
@@ -199,12 +199,29 @@ def _kl_binomial_binomial(p, q):
...
+def _kl_binomial_poisson(p, q):
+    _, (e1, _, e3) = p._Elnchoosek()
+    return (e1 - e3 +
+            p.mean * (p.logits - q.rate.log()) +
+            p.total_count * p._log1pmprobs() +
+            q.rate)
+
+
+@register_kl(Binomial, Geometric)
```
Comment (vishwakftw) [inline torch/distributions/kl.py]: Same as above comment.

File: torch/distributions/kl.py [hunk 8]
```diff
@@ -273,6 +290,11 @@ def _kl_geometric_geometric(p, q):
     return -p.entropy() - torch.log1p(-q.probs) / p.probs - q.logits
 
 
+@register_kl(Geometric, Binomial)
```
Comment (vishwakftw) [inline torch/distributions/kl.py]: Same as above comment.

**Comment by alicanb** [comment]
@vishwakftw thanks for the comments! One thing I want us to work out before wrapping this up is an approximation to E[logk!] for large n, I tried Stirling's but couldn't come up with a closed form. Any ideas?

**Comment by vishwakftw** [comment]
I think we have to make use of Stirling's inequality and the Taylor series to compute this. I guess the reason you are unable to come up with a closed form is because of the log (k) term.

I tried using them, and for got about 0.5% relative error.

File: torch/distributions/binomial.py (hunk 4 above)
Comment (alicanb) [reply torch/distributions/binomial.py]: weird, I tried using `flip` and it didn't work before- maybe I messed with arguments...

**Comment by vishwakftw** [comment]
Also, are you going to try the large `n` approximation using Stirling and Taylor expansions? @alicanb

**Comment by alicanb** [comment]
@vishwakftw btw I tried it with 0.01 precision as well. 2 things on my wishlist:
- large n approximation for `_Elnchoosek`
- KL(Bin(N,p)|Bin(M,p)) where M>N. Although we can calculate this expensively, making it work for batch is hard... Maybe it doesn't worth the effort.

**Comment by vishwakftw** [comment]
@alicanb I have a closed form solution for E[log x!], E[log (n - x)!] and E[log n!] (this is simply log n!) for large `n`.

**Comment by alicanb** [comment]
Great, have you experimented with any large `n`? `n`=30 seems not large enough for KL(Bin|Geom) for me with 0.1 precision.

**Comment by vishwakftw** [comment]
[This](https://gist.github.com/vishwakftw/8812a4abddca240c11737cd51d618108) is the gist for the approximations.

I ran some tests: n = {10, 20, 50, 75, 100} and p = {0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9}
Max relative error: 0.198 (n = 10, p = 0.9) and min relative error: 0.00025 (n = 100, p = 0.1). This is for E[log(n - x)!]

**Comment by alicanb** [comment]
btw `(lgamma(n * (1-p) + 1) + 0.5 * polygamma(1,n * (1-p) + 1) * n * p * (1-p))` is a pretty good approximation even for small `n`, but it's non-differentiable...
